executor/executor.c

```diff
@@ -472,18 +473,27 @@ int main(int argc, char* argv[]) {
         student_code = argv[1];
     }
     log_printf(DEBUG, "Min time %llu", min_time);
-    remove(CHALLENGE_SOCKET); // Always remove any old challenge socket
-    struct sockaddr_un my_addr = {AF_UNIX, CHALLENGE_SOCKET};    //for holding IP addresses (IPv4)
-	//create the socket
-	if ((challenge_fd = socket(AF_UNIX, SOCK_DGRAM, 0)) < 0) {
+	
+	// open challenge socket to read and write
+	struct sockaddr_un my_addr;
+	if ((challenge_fd = socket(AF_UNIX, SOCK_DGRAM, 0)) == -1) {
 		perror("socket");
-        log_printf(FATAL, "could not create challenge socket");
+		log_printf(FATAL, "could not create challenge socket");
 		return 1;
 	}
-
-	if (bind(challenge_fd, (struct sockaddr *) &my_addr, sizeof(struct sockaddr_un)) < 0) {
-        log_printf(FATAL, "challenge socket bind failed: %s", strerror(errno));
-		return 1;
+	//set up the challenge socket
+	memset(&my_addr, 0, sizeof(struct sockaddr_un));
+	my_addr.sun_family = AF_UNIX;
+	strcpy(my_addr.sun_path, CHALLENGE_SOCKET);
+	
+	//bind it if it doesn't exist on the system already
+	if (access(CHALLENGE_SOCKET, F_OK) == -1) {
+		int my_addr_len = offsetof(struct sockaddr_un, sun_path) + strlen(my_addr.sun_path);

```

Comment:
So this was in my networking book, the reason why there needs to be that is that the `struct sockaddr_un` structure is defined differently on different systems. Some contain extra fields that are strings of variable length, which means that `sizeof(sockaddr_un)` isn't actually going to give you the correct size of the `struct sockaddr_un`. That's why the `offsetof` is needed...

I was a bit confused about the challenge socket creation / removal though. I thought the challenge socket was kinda like a pipe, in that it's a file on the system that multiple processes can bind to and send/receive data to and from. If that's the case, it shouldn't be removed at all (kind of like how we don't remove `log-fifo` once we create it). 
That's why I have the logic in both `executor` and `net_handler` to basically make the socket if it doesn't exist, otherwise don't make it and just bind to it.

Reply:
This code is only run on 1 system, Linux on raspberry pi. On the Linux man page, the sockaddr_un seems defined to me and theres no mention of it being different: https://linux.die.net/man/7/unix. This is a case I would call unnecessarily complex. I'm not even sure why you removed the struct initializer I had before, because sockaddr_un has no other members so theres no need to initialize it to 0 first.

The reason that log-fifo isn't deleted is so that the logs on the pipe are saved. The challenge socket should be deleted by the executor because we want no one to have access to it if executor is not running. The challenge inputs should not be left in the socket cus it can block the next run of the challenge mode. The challenge sockets creation and removal should solely belong to the executor. I would also call this alternative unnecessarily complex...

Reply:
Ok, that makes sense. I'll change it back and hopefully everything works smoothly~~

Thanks for the explanation!